CalibTracker/SiPhase2TrackerESProducers/plugins/SiPhase2BadStripConfigurableFakeESSource.cc

```diff
@@ -0,0 +1,255 @@
+#include "CLHEP/Random/JamesRandom.h"
+
+class SiPhase2BadStripConfigurableFakeESSource : public edm::ESProducer, public edm::EventSetupRecordIntervalFinder {
+public:
+  SiPhase2BadStripConfigurableFakeESSource(const edm::ParameterSet&);
+  ~SiPhase2BadStripConfigurableFakeESSource() override = default;
+
+  void setIntervalFor(const edm::eventsetup::EventSetupRecordKey&,
+                      const edm::IOVSyncValue& iov,
+                      edm::ValidityInterval& iValidity) override;
+
+  typedef std::unique_ptr<SiStripBadStrip> ReturnType;
+  ReturnType produce(const SiPhase2OuterTrackerBadStripRcd&);
+
+  static void fillDescriptions(edm::ConfigurationDescriptions&);
+
+private:
+  std::map<unsigned short, unsigned short> clusterizeBadChannels(
+      const std::vector<Phase2TrackerDigi::PackedDigiType>& maskedChannels);
+
+  // configurables
+  bool m_printDebug;
+  float m_badComponentsFraction;
+
+  // random engine
+  std::unique_ptr<CLHEP::HepRandomEngine> m_engine;
+
+  // es tokens
+  edm::ESGetToken<TrackerTopology, TrackerTopologyRcd> m_trackTopoToken;
+  edm::ESGetToken<TrackerGeometry, TrackerDigiGeometryRecord> m_geomToken;
+};
+
+SiPhase2BadStripConfigurableFakeESSource::SiPhase2BadStripConfigurableFakeESSource(const edm::ParameterSet& iConfig)
+    : m_engine(new CLHEP::HepJamesRandom(iConfig.getParameter<unsigned int>("seed"))) {
+  auto cc = setWhatProduced(this);
+  m_trackTopoToken = cc.consumes();
+  m_geomToken = cc.consumes();
+
+  m_printDebug = iConfig.getUntrackedParameter<bool>("printDebug", false);
+  m_badComponentsFraction = iConfig.getParameter<double>("badComponentsFraction");
+
+  if (m_badComponentsFraction > 1. || m_badComponentsFraction < 0.) {
+    throw cms::Exception("Inconsistent configuration")
+        << "[SiPhase2BadStripChannelBuilder::c'tor] the requested fraction of bad components is unphysical. \n";
+  }
+
+  findingRecord<SiPhase2OuterTrackerBadStripRcd>();
+}
+
+void SiPhase2BadStripConfigurableFakeESSource::setIntervalFor(const edm::eventsetup::EventSetupRecordKey&,
+                                                              const edm::IOVSyncValue& iov,
+                                                              edm::ValidityInterval& iValidity) {
+  iValidity = edm::ValidityInterval{iov.beginOfTime(), iov.endOfTime()};
+}
+
+// ------------ method called to produce the data  ------------
+SiPhase2BadStripConfigurableFakeESSource::ReturnType SiPhase2BadStripConfigurableFakeESSource::produce(
+    const SiPhase2OuterTrackerBadStripRcd& iRecord) {
+  using namespace edm::es;
+  using Phase2TrackerGeomDetUnit = PixelGeomDetUnit;
+
+  //TrackerTopology const& tTopo = iRecord.get(m_trackTopoToken);
+  TrackerGeometry const& tGeom = iRecord.get(m_geomToken);
+
+  auto badStrips = std::make_unique<SiStripBadStrip>();
+
+  // early return with nullptr if fraction is ==0.
+  if (m_badComponentsFraction > 0.) {
+    ;  // do nothing
+  } else {
+    return badStrips;
+  }
```

Comment:
what is the advantage of doing this instead of just doing
```suggestion
  if (m_badComponentsFraction <= 0.) {
    return badStrips;
  }
```
?

Reply:
the case where it's strictly < 0 throws an exception at: 

https://github.com/mmusich/cmssw/blob/5151503f0f1ef6dd5e44bdcbf602e45482136755/CalibTracker/SiPhase2TrackerESProducers/plugins/SiPhase2BadStripConfigurableFakeESSource.cc#L82-L85

while for exact `m_badComponentsFraction == 0.f` floating point comparisons I've grown suspicious after the comments I have received at: https://github.com/cms-sw/cmssw/pull/37066#discussion_r816578741. If what you proposed is guaranteed to work I can change it.

Reply:
Ok, what about 
```suggestion
  if (!m_badComponentsFraction > 0.) {
    return badStrips;
  }
```
then?

Reply:
this is exactly what I **don't** want to do... :D it should early `return` only when it's exactly `==0.f`

Reply:
Ahh ok I understand, ok thanks for the clarification! Probably @perrotta would be better to comment on the
```
if (m_badComponentsFraction <= 0.) {
    return badStrips;
  }
```
solution to be guaranteed to be working.

Reply:
Floating point value `0.` is well defined, no other value compares equal to it. I suppose worry is if the value of `m_badComponentsFraction` would be a result of a computation, which is not the case here. Converting `0.` between `float` and `double` preserves the "zeroness".

Therefore
```cpp
if (m_badComponentsFraction > 0.) {} else { return badStrips; }

if (m_badComponentsFraction <= 0.) { return badStrips; }

if (m_badComponentsFraction == 0.) { return badStrips; } // because of the earlier exception

if (not m_badComponentsFraction > 0.) else { return badStrips; }
```
have identical behavior.

I would suggest though declaring `m_badComponentsFraction` as `double`, or using `float` literals (`0.f`, `1.f`) in the comparisons.

Reply:
decided to use `float` and compare with literals. Applied in the last push.